cmd/noms/noms_sync.go

```diff
@@ -77,10 +77,21 @@ func runSync(args []string) int {
 		lastProgressCh <- last
 	}()
 
+	sourceRef := types.NewRef(sourceObj)
+	sinkRef, _ := sinkDataset.MaybeHeadRef()
+	nonFF := false
 	err = d.Try(func() {
 		defer profile.MaybeStartProfile().Stop()
+		sinkDataset.Pull(sourceStore, sourceRef, p, progressCh)
+
 		var err error
-		sinkDataset, err = sinkDataset.Pull(sourceStore, types.NewRef(sourceObj), p, progressCh)
+		if sourceRef.Height() > sinkRef.Height() {
```

Comment:
Also, tucking the checks away inside Dataset API would make it hard to get out the info I need to print out the right user feedback. I suppose that whatever functions we changed/added could return some kind of indicator, but that would seem weird to me. It'd be returning a new Dataset, an error, and then some kind of return code as well to indicate what manner of head-setting it did.